### PR TITLE
HTTP/2 Lazy Children Map

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The children map in the DefaultStream class requires if checks before it is modified. This conditional checking is only applicable the first time the colleciton is access, but because of the structure current structure cannot be avoided.

Modifications:
- Add an abstraction interface ChildrenMap and add two implementations: LazyChildrenMap and DefaultChildrenMap. The LazyChildren map will transform itself into a DefaultChildrenMap upon first access.

Result:
Lazy children map with no conditional checks.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler @buchgr - FYI.

This was refactored to build on @buchgr's PR https://github.com/netty/netty/pull/3574.

**Comment by buchgr** [comment]
@Scottmitch 

This PR doesn't seem to make much sense to me? If I am understanding this correctly you are replacing one `if condition` with 100 additional lines of code and an additional object allocation?

**Comment by Scottmitch** [comment]
The tradeoff is less conditional checks (currently 1 conditional on every parent change event for every stream + more 1 more conditional check for each child if there is an exclusive dependency) for object allocation (1 additional object allocation the first time a parent change happens + 1 single extra if there is an exclusive dependency).  Conditionals create branch statements and can be problematic for optimizations (optimize it out if it evaluates to false very often...but then maybe put it back it eventually).  There is more code but at the expense of removing this conditional that is going to evaluate 1 way the majority of the time that I don't think can be optimized out because it behavior is different across threads.

**Comment by Scottmitch** [comment]
I know you have been looking at performance recently....do you have a way to evaluate this?  If not I can throw a micro benchmark together and see what happens.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 1]
```diff
@@ -615,6 +609,111 @@ final boolean removeChild(DefaultStream child) {
...
+        @Override
+        public Http2Stream remove(int streamId) {
+            return null;
+        }
+
+        @Override
+        public Http2Stream put(int id, DefaultStream stream) {
+            // Now that we are writing to the map we actually have to create it.
+            lazyStream.children = new DefaultChildMap();
+            return lazyStream.children.put(id, stream);
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: :)

**Comment by nmittler** [comment]
@Scottmitch LGTM

**Comment by buchgr** [comment]
@Scottmitch I am really sorry that I beat you to this PR. I didn't know you were working on this (maybe in the future just open an issue on all the stuff you are working on until it gets approved, or maybe in a group chat with @nmittler ?). But honestly, the whole point of my PR was to not allocate an additional object for every stream. This PR again allocates an object per stream. Yes, it is a bit smaller than the `IntObjectHashMap` (but not by much), but as we have seen with `Iterators` those small objects count.

So unless there is evidence that this conditional is a performance bottleneck, that justifies introducing 100 LOC and a object allocation per stream, I think this PR shouldn't be merged.

And yes, those "lazy objects" do account for some allocations (0.11% of the total heap in my benchmarks)
screenshot from 2015-04-03 15 36 30

**Comment by Scottmitch** [comment]
@buchgr - No need to apologize :) I agree there will be some additional memory overhead.  I'll try to come up with a benchmark and see if there is any significant difference in performance.  Won't move on this unless we can say more about if the performance benefit exists (may just be talking out my ass :)) and significant enough.  What is the memory consumption for the same scenario before your PR?

**Comment by buchgr** [comment]
@Scottmitch the call to `newChildMap()` was responsible for roughly 450MB in allocations. The `LazyPropertyMap` for 108MB. Number of object allocations is the same in both cases and that's what really counts ofc.

**Comment by Scottmitch** [comment]
@netkins build

**Comment by normanmaurer** [comment]
@Scottmitch @buchgr so what is the take away here now ? Just leave it as it is atm ?

**Comment by Scottmitch** [comment]
Lets kill this for now.  Did some simple benchmarking and I don't think there is any noticeable performance benefits.

**Comment by Scottmitch** [comment]
Thanks for the input @buchgr @nmittler ... sorry for the noise.

**Comment by normanmaurer** [comment]
@Scottmitch thanks buddy :)

**Pull request closed without merging**